// @azure-tools/typespec-azure-resource-manager/dist/src/rules/envelope-properties.js
import { createRule, getProperty } from "@typespec/compiler";
import { getArmResources } from "../resource.js";
import { getProperties } from "./utils.js";
/**
 * verify the 'identity' property should be present in the update resource properties.
 */
export const envelopePropertiesRules = createRule({
    name: "empty-updateable-properties",
    severity: "warning",
    description: "Should have updateable properties.",
    messages: {
        default: `The RP-specific properties of the Resource (as defined in the 'properties' property) should have at least one updateable property.  Properties are updateable if they do not have a '@visibility' decorator, or if they include 'update' in the '@visibility' decorator arguments.`,
    },
    create(context) {
        return {
            model: (model) => {
                var _a, _b, _c;
                const resources = getArmResources(context.program);
                const armResource = resources.find((re) => re.typespecType === model);
                if (armResource &&
                    armResource.operations.lifecycle.update &&
                    armResource.operations.lifecycle.createOrUpdate) {
                    const updateOperationProperties = (_a = armResource.operations.lifecycle.update.httpOperation.parameters.body) === null || _a === void 0 ? void 0 : _a.type;
                    if ((updateOperationProperties === null || updateOperationProperties === void 0 ? void 0 : updateOperationProperties.kind) === "Model") {
                        if (getProperties(updateOperationProperties).length < 1) {
                            context.reportDiagnostic({
                                target: model,
                            });
                        }
                        const updateablePropertiesBag = getProperty(updateOperationProperties, "properties");
                        if ((updateablePropertiesBag === null || updateablePropertiesBag === void 0 ? void 0 : updateablePropertiesBag.type.kind) === "Model" &&
                            getProperties(updateablePropertiesBag.type).length === 0) {
                            context.reportDiagnostic({
                                target: (_c = (_b = model.properties.get("properties")) === null || _b === void 0 ? void 0 : _b.type) !== null && _c !== void 0 ? _c : model,
                            });
                        }
                    }
                }
            },
        };
    },
});
