// @azure-tools/typespec-azure-resource-manager/dist/src/rules/patch-envelope-properties.js
import { createRule, getProperty, paramMessage } from "@typespec/compiler";
import { getArmResources } from "../resource.js";
export const patchEnvelopePropertiesRules = createRule({
    name: "patch-envelope",
    severity: "warning",
    description: "Patch envelope properties should match the resource properties.",
    messages: {
        default: paramMessage `The Resource PATCH request for resource '${"resourceName"}' is missing envelope properties:  [${"propertyName"}]. Since these properties are supported in the resource, they must also be updatable via PATCH.`,
    },
    create(context) {
        const patchEnvelopeProperties = ["identity", "managedBy", "plan", "sku", "tags"];
        return {
            model: (model) => {
                var _a;
                const resources = getArmResources(context.program);
                const armResource = resources.find((re) => re.typespecType === model);
                if (armResource &&
                    armResource.operations.lifecycle.update &&
                    armResource.operations.lifecycle.createOrUpdate) {
                    const updateOperationProperties = (_a = armResource.operations.lifecycle.update.httpOperation.parameters.body) === null || _a === void 0 ? void 0 : _a.type;
                    const missingProperties = [];
                    for (const property of patchEnvelopeProperties) {
                        if (!checkForPatchProperty(property, model, updateOperationProperties)) {
                            missingProperties.push(property);
                        }
                    }
                    if (missingProperties.length > 0) {
                        context.reportDiagnostic({
                            format: { resourceName: model.name, propertyName: missingProperties.join(", ") },
                            target: (updateOperationProperties === null || updateOperationProperties === void 0 ? void 0 : updateOperationProperties.kind) === "Model"
                                ? updateOperationProperties
                                : armResource.operations.lifecycle.update.operation,
                        });
                    }
                }
            },
        };
    },
});
function checkForPatchProperty(propertyName, resourceModel, patchModel) {
    return (getProperty(resourceModel, propertyName) === undefined ||
        ((patchModel === null || patchModel === void 0 ? void 0 : patchModel.kind) === "Model" && getProperty(patchModel, propertyName) !== undefined));
}
